[dy.namic] 지하철 노선도 1 ~ 3 단계 pr입니다.

kdy2000 opened the pull request.

Commit:
feat : 역/노선 관리 기능 추가

Commit:
feat : 구간관리 기능 추가

Commit:
feat : 구간관리 기능 추가 (완성)

Commit:
refactor : step3 수행 (DB와 연동)

Commit:
refactor : step3 수행 (DB와 연동)

Commit:
fix : section에서 station을 지울때, 종점을 제대로 지우지 못하는 버그 수정

Commit:
chore : code convention 맞춰줌

Commit:
fix : 잘못된 line을 생성했을때, 오류메시지는 나지만, db에 반영이 되버리는 버그 수정 (line 생성시 이름이 서로 같은 경우도 예외처리 추가)

findstar commented on build.gradle:
```diff
@@ -21,9 +21,17 @@ dependencies {
     // log
     implementation 'net.rakugakibox.spring.boot:logback-access-spring-boot-starter:2.7.1'
+    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+    implementation 'org.springframework.boot:spring-boot-starter-jdbc'
```
동일한 라인이 중복되고 아래에도 있네요.

findstar commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,94 @@
+        jdbcTemplate.update(sqlQuery, id);
+    }
+
+    public boolean hasDuplicateName(String name) {
```
이 메서드는 현재 외부에서 호출하는 곳이 없으니 private 으로 선언하는 게 더 나아보입니다.
그리고 이런 경우에는 isExist 와 같은 메서드명을 사용합니다.

findstar commented on src/main/java/subway/DuplicateException.java:
```diff
@@ -0,0 +1,15 @@
+
+import org.springframework.http.HttpStatus;
+
+public class DuplicateException extends RuntimeException {
```
개별 Exception 으로 분리하였네요 💯

findstar commented on src/main/java/subway/section/SectionDao.java:
```diff
@@ -0,0 +1,266 @@
+            throw new CannotConstructRightSectionsForLine("잘못된 section 저장입니다.");
+        }
+
+        /* 중간에 끼워 넣는경우 1 : upstation을 기준으로 기존 section을 삭제하고, 새롭게 연결해주어야 할 section을 추가 */
+        if(existSameUpstationId(section.getLineId(), section.getUpStationId())) {
+            Section oldSection = getSectionByUpStationId(section.getLineId(), section.getUpStationId());
+            addNewBackwardSection(section, oldSection);
+            deleteById(oldSection.getId());
+        }
+
+        /* 중간에 끼워 넣는 경우 2 : downstation을 기준으로 기존 section을 삭제하고, 새롭게 연결해주어야 할 section을 추가 */
+        if(existSameDownStationId(section.getLineId(), section.getDownStationId())) {
+            Section oldSection = getSectionByDownStationId(section.getLineId(), section.getDownStationId());
+            addNewForwardSection(section, oldSection);
+            deleteById(oldSection.getId());
+        }
+
+        /* 전달된 새로운 section을 추가 */
```
Dao 에서 너무 많은 역할을 처리하고 있지는 않은가요? 
분리해보면 좋겠습니다.

findstar commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,111 @@
+        Line line = lineDao.findById(id).orElseThrow(() -> new NotFoundException("존재하지 않는 line id 입니다."));
+
+        List<StationResponse> stationResponses = new ArrayList<>();
+        List<Long> stationIds = sectionDao.findSortedIdsByLineId(id);
+
+        for(Long stationId : stationIds){
+            Station station = stationDao.findById(stationId).orElseThrow(() -> new NotFoundException("존재하지 않는 line id 입니다."));
+            stationResponses.add(new StationResponse(
+                station.getId(), station.getName()
+            ));
+        }
```
SectionDao 에서 StationIds 를 가져와서 로직을 처리하고 있습니다.

findstar reviewed (changes requested):
안녕하세요 dy. 제임스입니다.
지하철 미션 진행하시느라 수고 많으셨습니다.
몇가지 코멘트 남겼으니 확인부탁드립니다.
질문은 언제든지 환영입니다. DM 부탁드려요 😄 
계속 화이팅입니다.!

Commit:
refactor : buil.gradle 파일에서 중복되는 line 제거

Commit:
refactor(LineDao) : hasDuplicateName() -> isExist()

Commit:
refactor : exception 클래스들을 하나의 package로 모았다.

Commit:
refactor : stationResponse에서 station객체를 받아 생성하도록 함

Commit:
refactor : stationDao 구현 방법 수정

1. RowMapper implements : inner class -> anonymous class
2. save() 메서드 : duplicate를 체크해주는 메서드를 삭제하고, 예외가 발생하면 catch해서 throw 해주도록 함.
3. findById 메서드 : Optional을 return 한뒤, 이 메서드를 호출한 곳에서 empty일때 exception 발생 -> Station을 return 하도록 하고, 존재하지 않을 경우 자체적으로 exception이 발생하도록 함.

Commit:
feat | refactor : stationService layer 추가

Commit:
refactor(StationDao) : query문을 StationQuery에서 제공해주도록 하였다.

Commit:
refactor(StationRequest) : StationRequest에서 station객체를 생성하여 반환하도록 변경했다.

Commit:
refactor(LineDao) : 구현방법 수정. LineQuery에서 query statement를 제공

Commit:
refactor(StationDao.save()) : try-catch로 중복 검사 -> isExist() 메서드로 중복검사

Commit:
refactor : stationResponse에서 Station 객체를 생성하지 않도록 함

Commit:
refactor : LineService, SectionService layer 추가, LineController에서 dao를 가지지 않도록 함. | lineDto, sectionDto 추가

Commit:
refactor : StationService, SectionService, LineService 에서 소유하는 dao, service를 private final 로 변경

Commit:
refactor : sectionDao에 있는 기능들을 sectionService로 이관하였다.

Commit:
refactor(SectionDao) : sectionRowMapper 구현 방식 변경

Commit:
refactor(SectionService) : SectionService에 존재하는 기능들을 domain으로 이동 (Section, SectionsInOneLine)

Commit:
refactor(SectionsInOneLine) : sectionsInOneLine에 존재하는 기능을 다른 domain객체로 분산

Commit:
feat(SectionsInOneLine) : service의 단위별로 transactional 추가

Commit:
chore(LineResponse) : 불필요한 코드 제거

kdy2000 commented:
section에서 station객체를 소유하도록 하고 싶었는데, 그럴경우 sectionDao에서, db row를 section으로 mapping해주는 객체를 작성할수가 없었습니다. 외부에서 뭔가 injection? 을 해주던가 해야할것 같은데, 구현방법도 문제지만, 그렇게 하는게 바람직한가도 의문이었습니다. 그래서 section객체에서 그냥 stationId를 가지도록 하였습니다.

findstar reviewed (approved):
안녕하세요 제임스입니다. 
코멘트 드린 피드백 잘 반영해주셨네요. 
PR 머지하겠습니다.
다음단계에서는 패키지를 좀 더 잘 정리하도록 고민해주세요
그럼 계속 화이팅입니다. 💯
